### PR TITLE
Suppress ld warning in macOS

### LINKED ISSUE #42282: Suppress ld warning in macOS
> Make sure that `git diff` result is empty and you've just pulled fresh master. Try cleaning up cmake cache. Just in case, official build instructions are published here: https://clickhouse.com/docs/en/development/build/

**Operating system**

macOS 12.3

**Cmake version**

```shell
% cmake --version    
cmake version 3.24.0
```

**Ninja version**

```shell
% ninja --version        
1.11.0
```

**Compiler name and version**

```shell
% clang --version
Homebrew clang version 14.0.6
Target: arm64-apple-darwin21.4.0
Thread model: posix
InstalledDir: /opt/homebrew/opt/llvm/bin
```

**Full cmake and/or ninja output**

```shell
....
ld: warning: direct access in function 'DB::(anonymous namespace)::ProtobufSerializerBuilder::buildFieldSerializer(std::__1::basic_string_view<char, std::__1::char_traits<char> >, std::__1::shared_ptr<DB::IDataType const> const&, google::protobuf::FieldDescriptor const&, bool, bool)' from file 'src/libdbms.a(ProtobufSerializer.cpp.o)' to global weak symbol 'typeinfo name for DB::DataTypeEnum<short>' from file 'src/libdbms.a(DataTypeEnum.cpp.o)' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in function 'DB::(anonymous namespace)::ProtobufSerializerBuilder::buildFieldSerializer(std::__1::basic_string_view<char, std::__1::char_traits<char> >, std::__1::shared_ptr<DB::IDataType const> const&, google::protobuf::FieldDescriptor const&, bool, bool)' from file 'src/libdbms.a(ProtobufSerializer.cpp.o)' to global weak symbol 'typeinfo name for DB::DataTypeEnum<short>' from file 'src/libdbms.a(DataTypeEnum.cpp.o)' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in function 'DB::(anonymous namespace)::ProtobufSerializerBuilder::buildFieldSerializer(std::__1::basic_string_view<char, std::__1::char_traits<char> >, std::__1::shared_ptr<DB::IDataType const> const&, google::protobuf::FieldDescriptor const&, bool, bool)' from file 'src/libdbms.a(ProtobufSerializer.cpp.o)' to global weak symbol 'typeinfo name for DB::DataTypeEnum<signed char>' from file 'src/libdbms.a(DataTypeEnum.cpp.o)' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
ld: warning: direct access in function 'DB::(anonymous namespace)::ProtobufSerializerBuilder::buildFieldSerializer(std::__1::basic_string_view<char, std::__1::char_traits<char> >, std::__1::shared_ptr<DB::IDataType const> const&, google::protobuf::FieldDescriptor const&, bool, bool)' from file 'src/libdbms.a(ProtobufSerializer.cpp.o)' to global weak symbol 'typeinfo name for DB::DataTypeEnum<signed char>' from file 'src/libdbms.a(DataTypeEnum.cpp.o)' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
```

Any ways to suppress them? It output tons of those warnings each time I run `ninja clickhouse`, which is quiet annoying.

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
Signed-off-by: Lloyd-Pottiger <yan1579196623@gmail.com>

### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

use llvm `l64.lld` in macOS suppress ld warnings, close #42282

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
I don't fully understand this change.

If the problem exists in MacOS, why it does not exist in Linux?

If we can use `-f visibility=hidden`, why don't we use it always?
- the answer is: allow to have more symbol names in the symbol table to have better displaying of stack traces in absence of the debug info; although the dynamic symbol table is unnecessary for this and the static symbol table should be enough.

And we cannot use `-f visibility=hidden` for building special integration tools like `clickhouse-odbc-bridge` and `clickhouse-library-bridge`.

The comment you have in the code does not explain:
- what was the problem;
- why it exists;
- why it cannot be solved in a more natural way;
- what's wrong with MacOS;
- what is `-f visibility=hidden` and why it helps to avoid the problem.

**Comment by Lloyd-Pottiger** [comment]
I found the ld we can in MacOS is default rather than LLVM's LLD
https://github.com/ClickHouse/ClickHouse/blob/61ed0a536437d2c90d559c526143ca8784b47cf8/cmake/tools.cmake#L66-L89

**Comment by alexey-milovidov** [comment]
Please fix this place to always prefer the usage of LLD.

**Comment by Lloyd-Pottiger** [comment]
> Please fix this place to always prefer the usage of LLD.

yeah, I have try to do that but get some error. Due to some urgent affair, I will come back to this problem next week.

**Comment by Lloyd-Pottiger** [comment]
> If the problem exists in MacOS, why it does not exist in Linux?

Now I can make sure that those warnings are only generated by macos default linker `ld`. If we use llvm `lld` in macos, these warnings are gone. 

```shell
$ lld --version
llvm lld is a generic driver
Invoke ld.lld (Unix), ld64.lld (macOS), lld-link (Windows), wasm-ld (WebAssembly) instead
```

Howerver, some flags are not supported by all of them, like

```shell
ld64.lld: error: unknown argument '--gdb-index'
ld64.lld: error: unknown argument '--build-id=sha1'
```

@alexey-milovidov any suggestions to handle this?

**Comment by rschu1ze** [comment]
Please also see #42450.

Issue #42282 shows a warning
```
ld: warning: direct access in function 'DB::(anonymous namespace)::ProtobufSerializerBuilder::buildFieldSerializer(std::__1::basic_string_view<char, std::__1::char_traits<char> >, std::__1::shared_ptr<DB::IDataType const> const&, google::protobuf::FieldDescriptor const&, bool, bool)' from file 'src/libdbms.a(ProtobufSerializer.cpp.o)' to global weak symbol 'typeinfo name for DB::DataTypeEnum<short>' from file 'src/libdbms.a(DataTypeEnum.cpp.o)' means the weak symbol cannot be overridden at runtime. This was likely caused by different translation units being compiled with different visibility settings.
```
which I think is best solved by passing `-fvisibility=...` when compiling the protobuf sources in ClickHouse (contrib/protobuf-cmake/CMakeLists.txt).

Besides that, I agree with Alexey: to minimize complexity in the build scripts (e.g. different supported arguments), we should prefer the Unix linker (ld) on Mac and support the build-in `ld64` only on a best effort basis (or not at all).

**Comment by Lloyd-Pottiger** [comment]
> which I think is best solved by passing `-fvisibility=...` when compiling the protobuf sources in ClickHouse (contrib/protobuf-cmake/CMakeLists.txt).
> 
> Besides that, I agree with Alexey: to minimize complexity in the build scripts (e.g. different supported arguments), we should prefer the Unix linker (ld) on Mac and support the build-in `ld64` only on a best effort basis (or not at all).

Sorry, I can not agree with you. For the master now, compiling in macos will output a lot of warnings involving most modules of CH but not just protobuf. And IMO, I prefer to use `ld64` in macos, but we only support llvm clang. Now the problem is, some flags are not supported by `ld64`, can we just remove it in macos?

**Comment by rschu1ze** [comment]
@Lloyd-Pottiger I can confirm that the default ld.lld produces the warnings listed in #42282 on Mac. The warnings go away with ld64.lld, i.e. (if I understand correctly) LLVM lld with Mach-O output which comes with brew's LLVM. As you write, this linker exposes the same flags as Apple's ld64.

I can imagine the linker warnings were introduced by the recent enablement of clang/lld/llvm 14 + 15 as compiler/linker but not sure.

For cross-compiling, we use Apple's ld64 and the linker path is also specified explicitly in the cmake invocation. So I don't think that this PR will interfere with cross-compiling.

If lld64.ld doesn't support some flags/options, it is okay to hide them behind `if (OS_DARWIN)` etc. in CMake. There may actually be lots of unsupported flags reachable only via `-D...=1` CMake option but IMHO it's sufficient if the [official build instructions](https://clickhouse.com/docs/en/development/build-osx/) work with ld64.lld. I am happy at least that `-fvisibility=` is off the table.

**Comment by Lloyd-Pottiger** [comment]
> @Lloyd-Pottiger I can confirm that the default ld.lld produces the warnings listed in #42282 on Mac. The warnings go away with ld64.lld, i.e. (if I understand correctly) LLVM lld with Mach-O output which comes with brew's LLVM. As you write, this linker exposes the same flags as Apple's ld64.
> 
> I can imagine the linker warnings were introduced by the recent enablement of clang/lld/llvm 14 + 15 as compiler/linker but not sure.
> 
> For cross-compiling, we use Apple's ld64 and the linker path is also specified explicitly in the cmake invocation. So I don't think that this PR will interfere with cross-compiling.
> 
> If lld64.ld doesn't support some flags/options, it is okay to hide them behind `if (OS_DARWIN)` etc. in CMake. There may actually be lots of unsupported flags reachable only via `-D...=1` CMake option but IMHO it's sufficient if the [official build instructions](https://clickhouse.com/docs/en/development/build-osx/) work with ld64.lld. I am happy at least that `-fvisibility=` is off the table.

lgtm, I think this pr is ready to be reviewed, could you please help to take a look?

**Review by rschu1ze — COMMENTED** [review]
Two minor comments, apart from that LGTM.

File: cmake/tools.cmake [hunk 1]
```diff
@@ -57,14 +57,19 @@ if (NOT LINKER_NAME)
...
         find_program (LLD_PATH NAMES "ld.lld")
         find_program (GOLD_PATH NAMES "ld.gold")
-    elseif (COMPILER_CLANG)
+    # llvm lld is a generic driver.
+    # Invoke ld.lld (Unix), ld64.lld (macOS), lld-link (Windows), wasm-ld (WebAssembly) instead
+    elseif (COMPILER_CLANG AND OS_LINUX)
         find_program (LLD_PATH NAMES "ld.lld-${COMPILER_VERSION_MAJOR}" "lld-${COMPILER_VERSION_MAJOR}" "ld.lld" "lld")
         find_program (GOLD_PATH NAMES "ld.gold" "gold")
+    elseif (COMPILER_CLANG AND OS_DARWIN)
+        find_program (LLD_PATH NAMES "ld64.lld-${COMPILER_VERSION_MAJOR}" "lld-${COMPILER_VERSION_MAJOR}" "ld64.lld" "lld")
```
[the remark this answers is not in the document]
Comment (Lloyd-Pottiger) [reply cmake/tools.cmake]: good idea, fixed!

**Review by rschu1ze — APPROVED** [review]
LGTM but let's wait for tests.

**Comment by xiedeyantu** [comment]
This pr make me run clickhouse throwing an exception: libunwind: stepWithCompactEncoding - invalid compact unwind encoding.

**Comment by alexey-milovidov** [comment]
@xiedeyantu download pre-built ClickHouse by running
```
curl https://clickhouse.com/ | sh
```

**Comment by chym1303** [comment]
@xiedeyantu I have encountered the same problem as you. After the PR, the debug compilation cannot run. After removing this PR, it can.

**Comment by xiedeyantu** [comment]
> @xiedeyantu I have encountered the same problem as you. After the PR, the debug compilation cannot run. After removing this PR, it can.

@Lloyd-Pottiger How to solve the problem?

**Comment by Lloyd-Pottiger** [comment]
> > @xiedeyantu I have encountered the same problem as you. After the PR, the debug compilation cannot run. After removing this PR, it can.
> 
> @Lloyd-Pottiger How to solve the problem?

What is your env(clang, CPU, os etc)? I did not meet the error on my local machine 🤔

**Comment by rschu1ze** [comment]
@xiedeyantu Please also paste your exact cmake invocation, thanks.

**Comment by rschu1ze** [comment]
Ups, now, I ran into the same issue on macOS ` libunwind: stepWithCompactEncoding - invalid compact unwind encoding` when the server starts, and reverting this PR fixes it. It looks like Apple introduced a new unwinding format ("compact" format) which is emitted by ld64.lld but not ld.lld (which was previously used) and libunwind (our own?) does not support it.

My CMake invocation:
```sh
$ cmake -DENABLE_EMBEDDED_COMPILER=0 -DENABLE_LIBRARIES=0 -DCMAKE_BUILD_TYPE=Debug -S . -B build
```

**Comment by Lloyd-Pottiger** [comment]
> Ups, now, I ran into the same issue on macOS `libunwind: stepWithCompactEncoding - invalid compact unwind encoding` when the server starts, and reverting this PR fixes it. It looks like Apple [introduced a new unwinding format ](https://faultlore.com/blah/compact-unwinding/)("compact" format) which is emitted by ld64.lld but not ld.lld (which was previously used) and libunwind (our own?) does not support it.
> 
> My CMake invocation:
> 
> ```shell
> $ cmake -DENABLE_EMBEDDED_COMPILER=0 -DENABLE_LIBRARIES=0 -DCMAKE_BUILD_TYPE=Debug -S . -B build
> ```

yep, you are right. 

https://github.com/ClickHouse/libunwind/blob/5022f30f3e092a54a7c101c335ce5e08769db366/src/CompactUnwinder.hpp#L516-L528

the libunwind we used just support `FRAME` and `FRAMELESS`, and do not support `DWARF`. But I also check llvm libunwind, it also does not support `DWARF`...

https://github.com/llvm/llvm-project/blob/6ee2f770efb6b1b02438db025af0899ed5bf9313/libunwind/src/CompactUnwinder.hpp#L517-L528

**Pull request merged**